### PR TITLE
Unimplemented features moved to their own exception

### LINKED ISSUE #1133: Split out UnimplementedFeature from InternalCompilerError exception
(Split out of #1125.)

Every error whose description contains something like "not yet implemented" should be put into its own error class `UnimplementedFeature` to distinguish it from `InternalCompilerError` especially to improve fuzz testing.

### PR DESCRIPTION
**Pull request by roadriverrail**
InternalCompilerError is an exception that really should be reserved for
actual internal errors of the compiler.  Unimplemented features can now
use either solUnimplemented( ) or, if it should be conditional, then
solUnimplementedAssert( ).

fixes #1133

### CODE REVIEW

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -1382,7 +1382,7 @@ void ExpressionCompiler::appendShiftOperatorCode(Token::Value _operator)
 	case Token::SHR:
 		break;
 	default:
-		BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("Unknown shift operator."));
+		BOOST_THROW_EXCEPTION(UnimplementedFeatureError() << errinfo_comment("Unknown shift operator."));
```
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp]: This is not an unimplemented feature.

Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp]: This is not an unimplemented feature.

**Review by axic — CHANGES_REQUESTED** [review]
Thanks for the PR!

A separate case must also be added to solc/jsonCompiler and solc/CommandLineInterface.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 2]
```diff
@@ -1243,7 +1243,7 @@ void ExpressionCompiler::endVisit(Literal const& _literal)
 	case Type::Category::StringLiteral:
 		break; // will be done during conversion
 	default:
-		BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("Only integer, boolean and string literals implemented for now."));
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:1246]: I think this should be a proper internal error.

**Review by chriseth — CHANGES_REQUESTED** [review]

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 1 above)
Comment (roadriverrail) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Wow.  I cannot properly explain myself for that one.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 2 above)
Comment (roadriverrail) [reply libsolidity/codegen/ExpressionCompiler.cpp:1246]: I was mostly going of the error text to make a guess, so no worries.

File: solc/CommandLineInterface.cpp [hunk 3]
```diff
@@ -604,6 +604,12 @@ bool CommandLineInterface::processInput()
 			 << boost::diagnostic_information(_exception);
 		return false;
 	}
+	catch (UnimplementedFeatureError const& _exception)
+	{
+		cerr << "Unimplemented compliler feature:" << endl
```
Comment (axic) [inline solc/CommandLineInterface.cpp]: Typo 😄  

Please keep the two messages consistent. I think I prefer "Unimplemented feature" over "Unimplemented compiler feature".

Comment (roadriverrail) [reply solc/CommandLineInterface.cpp]: Doh.

Happy to change them to "Unimplemented feature".  I think I was mostly keying off the language in the other messages.

**Review by chriseth — APPROVED** [review]

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
@roadriverrail thanks, merged!

**Pull request merged**